Cancel product

matthieu-rolland opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | Cancel product feature on migrated order detail page.
| Type?         | improvement
| Category?     | CO
| BC breaks?    | Does it break backward compatibility? yes/no
| Deprecations? | Does it deprecate an existing feature? yes/no
| Fixed ticket? | #16650
| How to test?  | Make an order in frontend, then go the order page in BO, click on "cancel product" at the top of the page, then select one or more product and save. New Order page URL:  **/admin-dev/sell/orders/orders/8/view** . (where 8 is the order ID)


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/17149)
<!-- Reviewable:end -->

matthieu-rolland committed:
cancel product feature

matthieu-rolland commented:
Not to be reviewed for now, this PR shares some js code with #17211 's javascript, there will be conflicts: waiting for #17211 to be approved, then small adaptations will be done and this PR will be reviewable.

jolelievre commented on admin-dev/themes/new-theme/js/pages/order/view/order-view-page.js:
```diff
@@ -229,16 +229,12 @@ export default class OrderViewPage {
       this.orderProductRenderer.moveProductsPanelToRefundPosition();
-    })
+    });
+    $(OrderViewPageMap.cancelProductSelectorCheckbox).on('change', (event) => {
```
This would be more logical to have this in the `order-product-cancel.js` file

jolelievre commented on src/Adapter/Order/CommandHandler/CancelOrderProductHandler.php:
```diff
@@ -0,0 +1,169 @@
+     */
+    public function handle(CancelOrderProductCommand $command)
+    {
+        $this->translator = Context::getContext()->getTranslator();
```
You should inject the translator rather than getting it from the context

jolelievre replied:
The translator still needs to be injected, not retrieved from the context

matthieu-rolland replied:
done

LouiseBonnard commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:
```diff
@@ -1203,12 +1245,41 @@ private function getErrorMessages(Exception $e)
+                    'Admin.Orderscustomers.Notification'
+                ),
+                InvalidCancelQuantityException::QUANTITY_TOO_HIGH => $this->trans(
+                    'An invalid quantity was selected for this product.',
```
If the exception is the same, do we have to make the string match the previous notification, _Please enter a maximum quantity of [1] to proceed with your refund._?

matthieu-rolland replied:
I can definitely do that 👍

LouiseBonnard commented on src/PrestaShopBundle/Form/Admin/Sell/Order/CancelProductType.php:
```diff
@@ -94,8 +122,38 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
+                    'multiple' => false,
+                    'expanded' => true,
+                    'choices' => [
+                        $this->trans('Product(s) price: ', 'Admin.Orderscustomers.Feature') => VoucherRefundType::PRODUCT_PRICES_REFUND,
```
One thought about it, @jolelievre: you told me that this string needs to have space at the end for UI purposes, right? My fear is that translators won't all be able to get the space at the end (when contributing on Crowdin) and reproduce it in their translation. As a consequence, it might give this page of the back office a bad interface.

LouiseBonnard commented on src/PrestaShopBundle/Form/Admin/Sell/Order/CancelProductType.php:
```diff
@@ -94,8 +122,38 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
+                    'expanded' => true,
+                    'choices' => [
+                        $this->trans('Product(s) price: ', 'Admin.Orderscustomers.Feature') => VoucherRefundType::PRODUCT_PRICES_REFUND,
+                        $this->trans('Product(s) price, excluding amount of initial voucher: ', 'Admin.Orderscustomers.Feature') => VoucherRefundType::PRODUCT_PRICES_EXCLUDING_VOUCHER_REFUND,
```
Same here, cf. previous comment.

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig:
```diff
@@ -32,85 +32,94 @@
+          <p>{{ 'Quantity'|trans({}, 'Admin.Global') }}</p>
+        </th>
+        <th>
+          <p>{{ 'Stock Location'|trans({}, 'Admin.Global') }}</p>
```
```suggestion
          <p>{{ 'Stock location'|trans({}, 'Admin.Orderscustomers.Feature') }}</p>
```

jolelievre replied:
Mmmm, careful I think this modif could be outdated or related to a wrong rebase
This part has been modified by the stock location PR https://github.com/PrestaShop/PrestaShop/pull/17135
And since the cancel product column has already been integrated in this template by the partial refund PR, I don't think you need to make any modification to this file

matthieu-rolland replied:
I got your html via the rebase I did 👍

jolelievre replied:
Considering the number of modification, it can't be the same as from the rebase ^^ But it doesn't matter it's only indentation modifications However you should make the modification asked by Louise

matthieu-rolland committed:
fix post rebase duplicate code

jolelievre commented on admin-dev/themes/new-theme/js/pages/order/OrderViewPageMap.js:
```diff
@@ -141,10 +141,12 @@ export default {
       save: '#cancel_product_save',
       partialRefund: 'button.partial-refund-display',
       standardRefund: 'button.standard-refund-display',
+      cancel: 'button.cancel-product-display',
```
```suggestion
      cancelProducts: 'button.cancel-product-display',
```
To avoid thinking it's a form cancel button

matthieu-rolland replied:
done

jolelievre commented on admin-dev/themes/new-theme/js/pages/order/OrderViewPageMap.js:
```diff
@@ -141,10 +141,12 @@ export default {
     inputs: {
       quantity: '.cancel-product-quantity input',
       amount: '.cancel-product-amount input',
+      selector: 'input[type=\'checkbox\'].selected-product',
```
There is already a similar class (matching quantity and amount selector) See https://github.com/PrestaShop/PrestaShop/pull/17416/files#diff-c8167039d3cf01caa7d474accb8bccf3R146

matthieu-rolland replied:
done

jolelievre commented on admin-dev/themes/new-theme/js/pages/order/view/order-product-cancel.js:
```diff
@@ -150,4 +150,29 @@ export default class OrderProductCancel {
+        $(OrderViewPageMap.cancelProduct.buttons.save).data('cancelLabel'),
+        cancelProductRoute
+    );
+    $(OrderViewPageMap.cancelProduct.form).attr('action', cancelProductRoute);
```
This is already managed by `initForm` So no need to do this twice

matthieu-rolland replied:
done

jolelievre commented on admin-dev/themes/new-theme/js/pages/order/OrderViewPageMap.js:
```diff
@@ -165,6 +167,10 @@ export default {
     toggle: {
       partialRefund: '.cancel-product-element:not(.hidden):not(.shipping-refund), .cancel-product-amount',
       standardRefund: '.cancel-product-element:not(.hidden):not(.shipping-refund), .cancel-product-selector',
+      cancel: '.cancel-product-element:not(.hidden):not(.shipping-refund), .cancel-product-selector',
```
Here as well you'd rather use `cancelProduct` or `cancelProducts` so we link it to the feature immediately, `cancel` alone is a bit too generic

matthieu-rolland replied:
done

jolelievre commented on admin-dev/themes/new-theme/js/pages/order/view/order-product-cancel.js:
```diff
@@ -150,4 +150,29 @@ export default class OrderProductCancel {
+        cancelProductRoute
+    );
+    $(OrderViewPageMap.cancelProduct.form).attr('action', cancelProductRoute);
+    this.hideRefund();
```
In the standard refund PR I introduced a `hideCancelElements` function to mutualize the elements hiding, this way we are sure not to forget some elements to hide since they are all hidden in the same function
https://github.com/PrestaShop/PrestaShop/pull/17416/files#diff-4ad12fe31814271b90489b62411921ddR89

matthieu-rolland replied:
Ok, to be fixed once your standard refund PR is merged.

matthieu-rolland replied:
done

jolelievre commented on admin-dev/themes/new-theme/js/pages/order/view/order-view-page.js:
```diff
@@ -263,6 +263,16 @@ export default class OrderViewPage {
+      this.orderProductCancel.showCancelProductForm();
+      this.orderProductRenderer.moveProductsPanelToRefundPosition();
+    });
+    $(OrderViewPageMap.cancelProduct.inputs.selector).on('change', (event) => {
```
This should be managed in the `orderProductCancel` component

matthieu-rolland replied:
done

jolelievre commented on admin-dev/themes/new-theme/js/pages/order/view/order-product-cancel.js:
```diff
@@ -150,4 +150,29 @@ export default class OrderProductCancel {
+    $(OrderViewPageMap.cancelProduct.containers.refundCheckboxes).hide();
+  }
+
+  fillCancelProductQuantityInput(productCheckbox) {
```
I added this as a more generic behavior (since it's also used by standard refund and return product) fully integrated in this class:
https://github.com/PrestaShop/PrestaShop/pull/17416/files#diff-4ad12fe31814271b90489b62411921ddR121

An "extra" behavior has also been decided after discussion with @MatShir and the QA team, we don't modify the quantity when it's different than 0

matthieu-rolland replied:
Ok, to be fixed once your standard refund PR is merged.

matthieu-rolland replied:
done

jolelievre commented on admin-dev/themes/new-theme/scss/pages/_orders.scss:
```diff
@@ -276,6 +262,9 @@ td .editProductInvoice {
         padding-left: 10px;
       }
     }
+    .cancel-product-quantity .input-group {
```
I made quite some modifications but I managed to integrate the inputs nicely in all features This involve changes in the css AND in the markup:
https://github.com/PrestaShop/PrestaShop/blob/f6119cb831fefa5b49be91e70e0c5d4deeadc569/admin-dev/themes/new-theme/scss/pages/_orders.scss
https://github.com/PrestaShop/PrestaShop/blob/f6119cb831fefa5b49be91e70e0c5d4deeadc569/src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig

matthieu-rolland replied:
Ok, to be fixed once your standard refund PR is merged.

matthieu-rolland replied:
yep I got your modifications while doing the rebase 👍

jolelievre replied:
Are you sure this one is still useful?

matthieu-rolland replied:
removed

sowbiba commented on src/Adapter/Order/CommandHandler/CancelOrderProductHandler.php:
```diff
@@ -0,0 +1,173 @@
+        $details = [];
+        $orderDetails = $this->getOrderDetails($command);
+
+        if (!empty($orderDetails['productsOrderDetails'])) {
```
No need to test empty I think. A foreach on an empty array just doesnt do anything

sowbiba commented on src/Adapter/Order/CommandHandler/CancelOrderProductHandler.php:
```diff
@@ -0,0 +1,173 @@
+            foreach ($orderDetails['productsOrderDetails'] as $orderDetail) {
+                $details[] = $orderDetail;
+                $customizationQuantity = 0;
+                $cancelQuantity = $orderDetails['productCancelQuantity'][$orderDetail->id_order_detail];
```
Maybe check that the key exists

sowbiba commented on src/Adapter/Order/CommandHandler/CancelOrderProductHandler.php:
```diff
@@ -0,0 +1,173 @@
+            foreach ($orderDetails['customizedProductsOrderDetail'] as $orderDetail) {
+                $customizationList[$orderDetail->id_customization] = $orderDetail->id_order_detail;
+            }
+            $customization_quantities = Customization::retrieveQuantitiesFromIds(array_keys($customizationList));
```
use camelCase for variables $customization_quantities $id_customization $id_order_detail, $customization_quantity, ...

matthieu-rolland replied:
ah yes, this whole class contains legacy code that was extracted from a legacy controller... I'll still address your feedback 👍

jolelievre replied:
If you use the `OrderRefundCalculator` I think most of this code becomes useless

jolelievre commented on src/Adapter/Order/CommandHandler/CancelOrderProductHandler.php:
```diff
@@ -0,0 +1,173 @@
+    /**
+     * Legacy code for product cancellation handling in order page
+     */
+    public function handle(CancelOrderProductCommand $command)
```
I think the whole handler code can be simplified through a few modifs:
- the invalid exceptions (negative quantity) should be managed in the command/valueobjects rather than in the handler
- you can use the the `OrderRefundCalculator` service to manage most calculations and parameters interpretation
- I'm not sure it's useful to separate the regular products and the customization in two separate arrays: you can simply loop through the OrderDetail list and check if `$orderDetail->id_customization` is present (see https://github.com/PrestaShop/PrestaShop/pull/17416/files#diff-61dbf27f8c77b84938556c403480bfe9R177 and https://github.com/PrestaShop/PrestaShop/pull/17416/files#diff-07c00c621d3ca10645633744ed48d14dR74)

sowbiba commented on src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php:
```diff
@@ -368,8 +368,10 @@ private function getOrderProducts(Order $order): OrderProductsForViewing
             $product['amount_refundable_tax_incl'] = $product['total_price_tax_incl'] - $product['total_refunded_tax_incl'];
-            $product['displayed_max_refundable'] = $taxCalculationMethod === PS_TAX_EXC ? $product['amount_refundable'] : $product['amount_refundable_tax_incl'];
-            $resumeAmountKey = $taxCalculationMethod === PS_TAX_EXC ? 'total_refunded_tax_excl' : 'total_refunded_tax_incl';
+            /* $product['displayed_max_refundable'] = $taxCalculationMethod === PS_TAX_EXC ? $product['amount_refundable'] : $product['amount_refundable_tax_incl'];
```
forgotten comments ?

matthieu-rolland replied:
yes, I would like the opinion of @jolelievre on this one (using PS_TAX_EXC vs getting the info from the order)

matthieu-rolland replied:
done

sowbiba commented on src/Core/Domain/Order/Command/CancelOrderProductCommand.php:
```diff
@@ -0,0 +1,87 @@
+     *
+     * key: orderDetailId, value: quantity
+     */
+    private $toBeCanceledProducts;
```
double L for cancelled

matthieu-rolland replied:
done

jolelievre commented on src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php:
```diff
@@ -368,8 +368,10 @@ private function getOrderProducts(Order $order): OrderProductsForViewing
-            $resumeAmountKey = $taxCalculationMethod === PS_TAX_EXC ? 'total_refunded_tax_excl' : 'total_refunded_tax_incl';
+            /* $product['displayed_max_refundable'] = $taxCalculationMethod === PS_TAX_EXC ? $product['amount_refundable'] : $product['amount_refundable_tax_incl'];
+            $resumeAmountKey = $taxCalculationMethod === PS_TAX_EXC ? 'total_refunded_tax_excl' : 'total_refunded_tax_incl'; */
+            $product['displayed_max_refundable'] = $order->getTaxCalculationMethod() ? $product['amount_refundable'] : $product['amount_refundable_tax_incl'];
```
This produces an error (in admin context this field is not correctly set in Order, so the amount was always displayed with taxes) that was fixed using `$taxCalculationMethod` (which is correctly set)

matthieu-rolland replied:
done

sowbiba commented on tests/Integration/Behaviour/Features/Context/Domain/OrderCancelProductFeatureContext.php:
```diff
@@ -0,0 +1,79 @@
+        $orderId = SharedStorage::getStorage()->get($orderReference);
+        $orderForViewing = $this->getQueryBus()->handle(new GetOrderForViewing((int) $orderId));
+        $products = $orderForViewing->getProducts()->getProducts();
+        $toBeCanceledProducts = [];
```
$toBeCancelledProducts

jolelievre commented on src/Core/Domain/Order/Command/CancelOrderProductCommand.php:
```diff
@@ -0,0 +1,87 @@
+     */
+    public function __construct(array $products, array $toBeCanceledProducts, OrderForViewing $order)
+    {
+        $this->products = $products;
```
I don't think you need the products list Only data about canceled products is enough You only use `getToBeCanceledProducts` in the handler, which seems logical since you can get the product list based on the orderId anyway

matthieu-rolland replied:
done

jolelievre commented on src/Core/Domain/Order/Command/CancelOrderProductCommand.php:
```diff
@@ -0,0 +1,87 @@
+     *
+     * key: orderDetailId, value: quantity
+     */
+    private $toBeCanceledProducts;
```
```suggestion
    private $canceledProducts;
```
For getters as well, it's simpler to read

matthieu-rolland replied:
done

jolelievre commented on src/Core/Domain/Order/Command/CancelOrderProductCommand.php:
```diff
@@ -0,0 +1,87 @@
+     * @param array $toBeCanceledProducts
+     * @param OrderForViewing $order
+     */
+    public function __construct(array $products, array $toBeCanceledProducts, OrderForViewing $order)
```
You are mixing `OrderForViewing` which is the result of a query in a command constructor
Command are supposed to use only scalar values, so you should use an `int $orderId` and store it in an `OrderId` value object

matthieu-rolland replied:
done

jolelievre commented on src/Core/Form/IdentifiableObject/DataHandler/CancellationFormDataHandler.php:
```diff
@@ -0,0 +1,88 @@
+     */
+    public function update($id, array $data)
+    {
+        $orderForViewing = $this->queryBus->handle(new GetOrderForViewing($id));
```
You don't need to request the order for viewing here, just use the orderId as is

matthieu-rolland replied:
done

jolelievre commented on src/Adapter/Order/CommandHandler/CancelOrderProductHandler.php:
```diff
@@ -0,0 +1,173 @@
+    public function handle(CancelOrderProductCommand $command)
+    {
+        $this->translator = Context::getContext()->getTranslator();
+        $order = new Order($command->getOrder()->getId());
```
You don't even use the `OrderForViewing` object, so it's sure it shouldn't be in the command

matthieu-rolland replied:
done

jolelievre commented on src/Core/Form/IdentifiableObject/DataProvider/CancelProductFormDataProvider.php:
```diff
@@ -33,7 +33,7 @@
 
 /**
- * Provides data for category add/edit category forms
+ * Provides data for partial refund form in order page
```
I'm not so sure ^^

matthieu-rolland replied:
done

jolelievre commented on src/Core/Form/IdentifiableObject/DataProvider/CancellationFormDataProvider.php:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Provides data for product cancellation form in order page
+ */
+final class CancellationFormDataProvider implements FormDataProviderInterface
```
I believe this class is not used any more since we have a common formtype/formdataprovider for all the features (same goes for related services)

matthieu-rolland replied:
done

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:
```diff
@@ -1174,6 +1176,26 @@ public function sendProcessOrderEmailAction(Request $request): JsonResponse
+            $result = $formHandler->handleFor($orderId, $form);
+            if ($result->isSubmitted() && $result->isValid()) {
+                $this->addFlash('success', $this->trans('The discount was successfully generated.', 'Admin.Catalog.Notification'));
+            }
```
You should display the errors here in case the form is not valid, I had the issue in some cases, see https://github.com/PrestaShop/PrestaShop/pull/17416/files#diff-3b09571065f487792fa092444ada455aR508

I believe it's not done in many of our controllers but we probably forgot it, or simply never had these kind of errors (in my case it was a csrf token error)

matthieu-rolland replied:
Ok, to be fixed once your standard refund PR is merged.

matthieu-rolland replied:
done

jolelievre commented on src/PrestaShopBundle/Form/Admin/Sell/Order/CancelProductType.php:
```diff
@@ -147,10 +148,11 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
             ->add('save', SubmitType::class, [
                 'attr' => [
-                    'class' => 'cancel-product-element save btn btn-primary ml-3',
+                    'class' => 'partial-refund save btn btn-primary ml-3',
```
Nope, wrong rebase or outdated code I think This calss is not used any more, the javascript relies on the `cancel-product-element` class for its toggling selectors

matthieu-rolland replied:
done

jolelievre commented on src/PrestaShopBundle/Form/Admin/Sell/Order/CancellationType.php:
```diff
@@ -0,0 +1,67 @@
+use Symfony\Component\Form\Extension\Core\Type\SubmitType;
+use Symfony\Component\Form\FormBuilderInterface;
+
+class CancellationType extends TranslatorAwareType
```
This form type is not used any more I think

matthieu-rolland replied:
done

jolelievre commented on src/PrestaShopBundle/Resources/config/services/adapter/order.yml:
```diff
@@ -166,3 +166,11 @@ services:
+  prestashop.adapter.order.command_handler.cancel_order_product_handler:
+    class: PrestaShop\PrestaShop\Adapter\Order\CommandHandler\CancelOrderProductHandler
+    arguments:
+      - "@prestashop.core.localization.locale.context_locale"
```
The arguments don't match the constructor

matthieu-rolland replied:
done

jolelievre commented on src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml:
```diff
@@ -1121,3 +1121,10 @@ services:
       tags:
         - { name: form.type }
+
+    form.type.order.cancellation:
```
Not used any more

matthieu-rolland replied:
done

jolelievre commented on src/PrestaShopBundle/Resources/config/services/core/form/form_builder.yml:
```diff
@@ -168,3 +168,13 @@ services:
       - 'PrestaShopBundle\Form\Admin\Sell\Address\CustomerAddressType'
       - '@prestashop.core.form.identifiable_object.data_provider.address_form_data_provider'
+
+  prestashop.core.form.identifiable_object.builder.cancellation_form_builder:
```
Not used any more

matthieu-rolland replied:
done

jolelievre commented on src/PrestaShopBundle/Resources/config/services/core/form/form_data_provider.yml:
```diff
@@ -138,3 +138,8 @@ services:
       - '@prestashop.core.query_bus'
       - '@=service("prestashop.adapter.legacy.context").getContext().country.id'
+
+  prestashop.core.form.identifiable_object.data_provider.cancellation_form_data_provider:
```
Not used any more

matthieu-rolland replied:
done

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/order_actions.html.twig:
```diff
@@ -94,5 +94,13 @@
       </div>
     </div>
+  {% elseif merchandiseReturnEnabled %}
+    <div class="form-inline d-inline-block form-horizontal ml-2">
```
This button should be integrated in the block a bit above `{% if merchandiseReturnEnabled %}` along with all the buttons related to the merchandise return feature, see https://github.com/PrestaShop/PrestaShop/pull/17416/files#diff-9ec92f94e56724fb3be768b86bcd689eR76

matthieu-rolland replied:
Ok, to be fixed once your standard refund PR is merged.

matthieu-rolland replied:
done

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/product.html.twig:
```diff
@@ -23,7 +23,7 @@
  *#}
 {% set rowIsDisplayed =  (productIndex is defined and paginationNum is defined and productIndex > paginationNum) %}
-<tr id="orderProduct_{{ product.orderDetailId }}"{% if rowIsDisplayed %} class="d-none"{% endif %}>
+<tr id="orderProduct_{{ product.orderDetailId }}" data-product-id="{{ product.orderDetailId }}" data-product-quantity="{{ product.quantity }}" {% if rowIsDisplayed %} class="d-none"{% endif %}>
```
I chose to integrate all the data required in javascript in the quantity input field, since it's the only that is common to all features it allows to have a consistent management

However, although I think it's clearer to have all the data on the same DOM element, storing this data in a more "parent" element makes sense (if one day we get rid or toggle the quantity input it could be bothersome), but then maybe in an element more related to the feature Either the cancel product `td`, or the `.cancel-product-cell-elements` div (see https://github.com/PrestaShop/PrestaShop/pull/17416/files#diff-6a8b04b47a9ab83236551aa8ec626e19R108)

matthieu-rolland replied:
ok, to be done once your standard refund PR is merged.

matthieu-rolland replied:
done

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/OrderCancelProductFeatureContext.php:
```diff
@@ -0,0 +1,79 @@
+     * @param string $orderReference
+     * @param TableNode $table
+     */
+    public function cancelOrderProduct(string $orderReference, TableNode $table)
```
I think you can integrate this in the `OrderRefundContext` since the four features are **very** similar it makes sense to share their testing classes/methods

matthieu-rolland replied:
done

jolelievre commented on tests/Integration/Behaviour/Features/Scenario/Order/order_cancel_product.feature:
```diff
@@ -0,0 +1,58 @@
+      | Mug Today is a good day     | 1        |
+    Then order "bo_order_cancel_product" should contain 3 products "Mug The best is yet to come"
+    And order "bo_order_cancel_product" should contain 2 products "Mug Today is a good day"
+    And there are 2 more "Mug The best is yet to come" in stock
```
Maybe you should add tests regarding what the OrderDetail should contain, does it update `product_quantity` ? Or does it increment `product_quantity_return` or `product_quantity_refunded`

What about `product_quantity_reinjected`? Since you call `reinjectQuantity` it's supposed to be modified as well

matthieu-rolland replied:
done

jolelievre commented on tests/Integration/Behaviour/Features/Scenario/Order/order_cancel_product.feature:
```diff
@@ -0,0 +1,58 @@
+    And there are 1 more "Mug Today is a good day" in stock
+
+  @order-cancel-product
+  Scenario: Quantity is required
```
Can you test negative values as well please?

matthieu-rolland replied:
done

jolelievre commented on tests/Integration/Behaviour/Features/Scenario/Order/order_cancel_product.feature:
```diff
@@ -0,0 +1,58 @@
+    And I add 3 products "Mug Today is a good day" to the cart "dummy_cart"
+    And I watch the stock of product "Mug The best is yet to come"
+    And I watch the stock of product "Mug Today is a good day"
+
```
You should add some scenarii tests related to the order state (of course you need to manage it in the handler as well)
I think cancel product is forbidden has soon as the order has an invoice?
Her's an example https://github.com/PrestaShop/PrestaShop/pull/17416/files#diff-0368d4838ab98bab3f54376d260b7136R554

matthieu-rolland replied:
yes, and also checking the canceled status of the order (when everything has been canceled)

matthieu-rolland replied:
done, I also checked that order state is set to cancelled after all products have been cancelled:

https://github.com/PrestaShop/PrestaShop/pull/17149/commits/f09de8b66abe7f83d6442d4df2dfb4fd28827bca

jolelievre reviewed (changes requested):
Sorry sorry... I struck again 😅

matthieu-rolland committed:
address a few feedbacks

matthieu-rolland committed:
cs fixer

matthieu-rolland committed:
move product data attribute and its usages

matthieu-rolland committed:
update products.html.twig

matthieu-rolland committed:
remove fillCancelProductQuantityInput method

matthieu-rolland committed:
Make OrderRefundException more generic

matthieu-rolland committed:
refactor invalid cancel quantity exception

matthieu-rolland committed:
test negative values (behat)

matthieu-rolland committed:
improve cancel product behat exceptions

jolelievre commented on admin-dev/themes/new-theme/js/pages/order/OrderViewPageMap.js:
```diff
@@ -169,6 +171,10 @@ export default {
       partialRefund: '.cancel-product-element:not(.hidden):not(.shipping-refund), .cancel-product-amount',
       standardRefund: '.cancel-product-element:not(.hidden):not(.shipping-refund-amount):not(.restock-products), .cancel-product-selector',
       returnProduct: '.cancel-product-element:not(.hidden):not(.shipping-refund-amount), .cancel-product-selector',
+      cancelProducts: '.cancel-product-element:not(.hidden):not(.shipping-refund), .cancel-product-selector',
```
Shouldn't you ignore the restock-product checkbox and voucher as well? (actually maybe the whole `refund-checkboxes-container` ? what is left?)

matthieu-rolland replied:
I'm not sure I understand what you mean ?

jolelievre replied:
This selector is used to show only the elements related to cancel products, so basically it shows all `.cancel-product-element` but with some exceptions because some feature don't need them all

For example partial refund doesn't need the `.shipping-refund` checkbox because it uses `.shipping-refund-amount` input, but for standard refund it's the opposite

Anyway this selector only should be enough to correctly shows the expected inputs, but in the previous js you used `refundCheckboxes` to hide these elements (now the selector is not used any more).

So tldr; I believe you don't need the restock checkbox for cancel product so you should add `:not(.restock-products)` And you don't need to generate voucher or credit slip either so you can hide them using the same principle (you will have to add the appropriate classes on the div containing them in the template)

matthieu-rolland replied:
oh ok I get it know, done 👍  I had to add two missing classes for this to work

matthieu-rolland committed:
deal with form errors for product cancellation

jolelievre commented on admin-dev/themes/new-theme/js/pages/order/view/order-product-cancel.js:
```diff
@@ -209,4 +209,16 @@ export default class OrderProductCancel {
+    );
+    this.hideCancelElements();
+    $(OrderViewPageMap.cancelProduct.toggle.cancelProducts).show();
+    $(OrderViewPageMap.cancelProduct.table.actions).hide();
```
It's already present in `hideCancelElements `

matthieu-rolland replied:
done

jolelievre commented on admin-dev/themes/new-theme/js/pages/order/view/order-product-cancel.js:
```diff
@@ -209,4 +209,16 @@ export default class OrderProductCancel {
+    this.hideCancelElements();
+    $(OrderViewPageMap.cancelProduct.toggle.cancelProducts).show();
+    $(OrderViewPageMap.cancelProduct.table.actions).hide();
+    $(OrderViewPageMap.cancelProduct.containers.refundCheckboxes).hide();
```
This should be managed via the `OrderViewPageMap.cancelProduct.toggle.cancelProducts` selector using `:not` on element you need to keep hidden

matthieu-rolland replied:
outdated

jolelievre commented on admin-dev/themes/new-theme/js/pages/order/view/order-product-cancel.js:
```diff
@@ -209,4 +209,16 @@ export default class OrderProductCancel {
+    const cancelProductRoute = this.router.generate('admin_orders_cancellation', {orderId: this.orderId});
+    this.initForm(
+        $(OrderViewPageMap.cancelProduct.buttons.save).data('cancelLabel'),
+        cancelProductRoute
```
There is a missing parameter, it is the class applied on the form, it's probably useless in this case (I think only partial-refund use it in CSS) But let's keep this consistent in case a special case is one day needed for cancel product display

matthieu-rolland replied:
done

jolelievre commented on admin-dev/themes/new-theme/js/pages/order/view/order-view-page.js:
```diff
@@ -268,6 +268,13 @@ export default class OrderViewPage {
 
+  listenForCancelProduct() {
+    $(OrderViewPageMap.cancelProduct.buttons.cancelProducts).on('click', (event) => {
+      this.orderProductCancel.showCancelProductForm();
```
Swap both lines please (to be consistent, and just in case some elements might be toggled in `moveProductsPanelToRefundPosition` and mess with the toggling in `showCancelProductForm`)

matthieu-rolland replied:
done

jolelievre commented on admin-dev/themes/new-theme/scss/pages/_orders.scss:
```diff
@@ -241,6 +241,14 @@ td.cancel-product-element {
     }
   }
+
+  .cancel-product-cell {
```
This shouldn't be needed any more I made thorough tests to check the display is correct everywhere Is this some remaining code from the beginning of the PR or do you really need this?

matthieu-rolland replied:
done

jolelievre commented on src/Core/Domain/Order/CommandHandler/CancelOrderProductHandlerInterface.php:
```diff
@@ -0,0 +1,40 @@
+use PrestaShop\PrestaShop\Core\Domain\Order\Command\CancelOrderProductCommand;
+
+/**
+ * Interface for service that handles issuing partial refund for given order
```
```suggestion
 * Interface for service that handles issuing cancel product for given order
```

jolelievre commented on src/Core/Domain/Order/Exception/EmptyCancelQuantityException.php:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Throw new when a product cancellation is asked without a specified quantity
+ */
+class EmptyCancelQuantityException extends OrderException
```
This one is useless now I guess (everything is managed by the same exception no?)

matthieu-rolland replied:
done

jolelievre commented on src/Core/Domain/Order/Exception/EmptyProductSelectionException.php:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Thrown when no product was selected
+ */
+class EmptyProductSelectionException extends OrderException
```
This one is useless now I guess (everything is managed by the same exception no?)

matthieu-rolland replied:
This one is used to check that at least one product has been selected (it's not the same as the InvalidCancelException, that checks a specific's product cancellation/refund etc (quantity/amount/refunds...).

I think it should have its own exception... it is thrown following a check that is done on all the form (not a specific product).

matthieu-rolland replied:
done

jolelievre commented on src/Adapter/Order/CommandHandler/CancelOrderProductHandler.php:
```diff
@@ -0,0 +1,196 @@
+    private function checkInput(CancelOrderProductCommand $command)
+    {
+        if (empty($command->getCancelledProducts())) {
+            throw new EmptyProductSelectionException();
```
This should be managed with `InvalidCancelProductException::NO_REFUNDS` (or maybe a new dedicated code) Bt no need for a dedicated exception

matthieu-rolland replied:
hm, answered above without reading this first, I didn't get that NO_REFUNDS was used in this case.

matthieu-rolland replied:
done

matthieu-rolland committed:
clean action button conditions

matthieu-rolland committed:
refactor cancel product feature context methods into refund feature context

matthieu-rolland committed:
test that order is cancelled when all products have been cancelled

jolelievre commented on admin-dev/themes/new-theme/js/pages/order/OrderViewPageMap.js:
```diff
@@ -152,6 +153,7 @@ export default {
       cell: '.cancel-product-cell',
       header: 'th.cancel-product-element p',
       actions: 'td.cellProductActions, th.product_actions',
+      cellElements: '.cancel-product-cell-elements',
```
I don't think you need this selector, it's not used anywhere

matthieu-rolland replied:
done

jolelievre replied:
Still here though ^^

matthieu-rolland replied:
ah, must have missed this file while commiting.

jolelievre commented on src/Core/Form/IdentifiableObject/DataHandler/CancellationFormDataHandler.php:
```diff
@@ -0,0 +1,84 @@
+     */
+    public function update($id, array $data)
+    {
+        $toBeCanceledProducts = [];
```
```suggestion
        $cancelledProducts = [];
```

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:
```diff
@@ -1314,6 +1315,28 @@ public function sendProcessOrderEmailAction(Request $request): JsonResponse
         }
     }
 
+    public function cancellationAction(int $orderId, Request $request)
```
Please add the security annotation

matthieu-rolland replied:
done

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:
```diff
@@ -1314,6 +1315,28 @@ public function sendProcessOrderEmailAction(Request $request): JsonResponse
+            if ($result->isSubmitted() && $result->isValid()) {
+                $this->addFlash('success', $this->trans('The discount was successfully generated.', 'Admin.Catalog.Notification'));
+            } else {
+                $this->addFlashFormErrors($form);
```
`addFlashFormErrors` must be called only when `$result->isSubmitted()` is true, check this controller https://github.com/PrestaShop/PrestaShop/pull/17416/files#diff-3b09571065f487792fa092444ada455aR511

The if statements are not exactly like this

matthieu-rolland replied:
done

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:
```diff
@@ -1365,29 +1388,33 @@ private function getErrorMessages(Exception $e)
-                ),
             ],
+            InvalidCancelProductException::INVALID_AMOUNT => $this->trans(
+                'Please enter a positive amount to proceed with your refund.',
```
I guess you can remove the end to be consistent with other messages (even though it's only used for a refund)

matthieu-rolland replied:
done

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:
```diff
@@ -1396,6 +1423,12 @@ private function getErrorMessages(Exception $e)
                 'Admin.Notifications.Error'
             ),
+            InvalidOrderStateException::class => [
+                InvalidOrderStateException::HAS_INVOICE => $this->trans(
```
Can you please add messages for other `InvalidOrderStateException` use cases

matthieu-rolland replied:
done

jolelievre commented on src/PrestaShopBundle/Form/Admin/Sell/Order/CancelProductType.php:
```diff
@@ -50,6 +50,7 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
                         'label' => false,
                         'attr' => [
                             'material_design' => true,
+                            'class' => 'selected-product',
```
Useless now

matthieu-rolland replied:
done

jolelievre commented on src/PrestaShopBundle/Resources/config/services/core/form/form_builder.yml:
```diff
@@ -168,3 +168,6 @@ services:
     arguments:
       - 'PrestaShopBundle\Form\Admin\Sell\Address\CustomerAddressType'
       - '@prestashop.core.form.identifiable_object.data_provider.address_form_data_provider'
+
```
Remove extra lines

matthieu-rolland replied:
done

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/order_actions.html.twig:
```diff
@@ -92,16 +92,22 @@
+        </div>
       </div>
-    </div>
+    {% else %}
```
Nope the partial refund button is completely independent from the merchandise return feature You musn't include it in the `if merchandiseReturnEnabled` block or you will remove the feature along with merchandise return

matthieu-rolland replied:
done

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/product.html.twig:
```diff
@@ -23,7 +23,7 @@
  *#}
 {% set rowIsDisplayed =  (productIndex is defined and paginationNum is defined and productIndex > paginationNum) %}
-<tr id="orderProduct_{{ product.orderDetailId }}"{% if rowIsDisplayed %} class="d-none"{% endif %}>
+<tr id="orderProduct_{{ product.orderDetailId }}" {% if rowIsDisplayed %} class="d-none"{% endif %}>
```
No need for extra space, the space is included in the condition on purpose (to avoid adding an extra useless space when it's false)

matthieu-rolland replied:
done

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/product.html.twig:
```diff
@@ -105,7 +105,7 @@
   <td class="text-right cancel-product-element">
     <div class="cancel-product-cell cancel-product-element{% if product.refundable == false %} hidden{% endif %}">
-      <div class="cancel-product-cell-elements">
+      <div class="cancel-product-cell-elements" data-product-id="{{ product.orderDetailId }}" data-product-quantity="{{ product.quantity }}">
```
The data on this element is not used It is all assigned on the quantity input

So you can either move all the data from the input to this div (preferably on the `cancel-product-cell` div even - I know I told you to use `cancel-product-cell-elements` in the previous comment ^^ I can change my mind) and update the js accordingly
Or you can leave the data on the input and js as is, since it works fine, and remove these ones

matthieu-rolland replied:
done

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig:
```diff
@@ -50,23 +50,34 @@
+          <p>{{ 'Refunded'|trans({}, 'Admin.Orderscustomers.Feature') }}</p>
+        </th>
+        <th>
+          <p>{{ 'Available'|trans({}, 'Admin.Global') }}</p>
```
WARNING!! you should revert all the changes on this file The rebase messed it up, some columns are duplicated The indentation is a bit better (sometimes) but nothing valuable is added here All the necessary columns are already present and well formatted in the target branch

matthieu-rolland replied:
done

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/OrderRefundFeatureContext.php:
```diff
@@ -223,45 +224,57 @@ public function enabledReturnProduct()
     }
 
+    /**
+     * @then I should get error that cancel quantity is too high
```
Why did you need to create a new endpoint You could have used the existing one which also checks that `getRefundableQuantity` is correctly set?

matthieu-rolland replied:
done

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/OrderRefundFeatureContext.php:
```diff
@@ -448,4 +461,37 @@ private function createIssueReturnProductCommand(
     }
+
+    /**
+     * @When I do a cancel from order :orderReference on the following products:
```
```suggestion
     * @When I cancel the following products from order :orderReference:
```

matthieu-rolland replied:
done

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/OrderRefundFeatureContext.php:
```diff
@@ -448,4 +461,37 @@ private function createIssueReturnProductCommand(
+        $orderId = SharedStorage::getStorage()->get($orderReference);
+        $orderForViewing = $this->getQueryBus()->handle(new GetOrderForViewing((int) $orderId));
+        $products = $orderForViewing->getProducts()->getProducts();
+        $toBeCanceledProducts = [];
```
```suggestion
        $cancelledProducts = [];
```

matthieu-rolland replied:
done

jolelievre commented on tests/Integration/Behaviour/Features/Scenario/Order/order_cancel_product.feature:
```diff
@@ -0,0 +1,124 @@
+    Then I should get error that cancel quantity is invalid
+
+  @order-cancel-product
+  Scenario: Quantity must be positive
```
```suggestion
  Scenario: Quantity is too high
```

jolelievre replied:
It's not the right scenario description

jolelievre replied:
Or not the right test, your choice ^^

matthieu-rolland replied:
done

jolelievre commented on tests/Integration/Behaviour/Features/Scenario/Order/order_cancel_product.feature:
```diff
@@ -0,0 +1,124 @@
+      | product_name                | quantity |
+      | Mug The best is yet to come | 1       |
+      | Mug Today is a good day     | 1        |
+    Then I should get error that order state is invalid
```
You should update this check, either create an endpoint for each possible exception code Or add a variable in the sentence to specify the excepted code

matthieu-rolland replied:
done

jolelievre commented on src/Core/Domain/Order/Exception/InvalidOrderStateException.php:
```diff
@@ -32,4 +32,33 @@
+    /**
+     * Used when the order has an invoice (and it should not)
+     */
+    const HAS_INVOICE = 2;
```
```suggestion
    const UNEXPECTED_INVOICE = 2;
```

jolelievre commented on src/Core/Domain/Order/Exception/InvalidOrderStateException.php:
```diff
@@ -32,4 +32,33 @@
+    /**
+     * Used when the order has no invoice (and it should have)
+     */
+    const HAS_NO_INVOICE = 1;
```
```suggestion
    const INVOICE_NOT_FOUND = 1;
```

jolelievre commented on src/Core/Domain/Order/Exception/InvalidOrderStateException.php:
```diff
@@ -32,4 +32,33 @@
+    /**
+     * Used when the order has been delivered (and it shouldn't have)
+     */
+    const HAS_BEEN_DELIVERED = 4;
```
```suggestion
    const UNEXPECTED_DELIVERY = 4;
```

jolelievre commented on src/Core/Domain/Order/Exception/InvalidOrderStateException.php:
```diff
@@ -32,4 +32,33 @@
+    /**
+     * Used when the order has not been delivered (and it should have)
+     */
+    const HAS_NOT_BEEN_DELIVERED = 3;
```
```suggestion
    const DELIVERY_NOT_FOUND = 3;
```

matthieu-rolland committed:
cs fixer

jolelievre reviewed (changes requested):
Some new feedback ^^
Also good job on the `InvalidOrderStateException` but you need to update the other handlers using it so that they use the appropriate code, and then update the behat tests to check the code
For `IssueStandardRefundHandler` indeed it has two invalid states, but since each one already has the corresponding code you can adapt which one needs to be thrown (and I each one already has a dedicated behat test so you can adapt them by checking the correct code)

matthieu-rolland committed:
address a few feedbacks

matthieu-rolland committed:
Update src/Core/Domain/Order/CommandHandler/CancelOrderProductHandlerInterface.php

Co-Authored-By: jolelievre <jo.lelievre@gmail.com>

matthieu-rolland committed:
Update src/Core/Form/IdentifiableObject/DataHandler/CancellationFormDataHandler.php

Co-Authored-By: jolelievre <jo.lelievre@gmail.com>

matthieu-rolland committed:
Update src/Core/Domain/Order/Exception/InvalidOrderStateException.php

Co-Authored-By: jolelievre <jo.lelievre@gmail.com>

matthieu-rolland committed:
Update src/Core/Domain/Order/Exception/InvalidOrderStateException.php

Co-Authored-By: jolelievre <jo.lelievre@gmail.com>

matthieu-rolland committed:
Update src/Core/Domain/Order/Exception/InvalidOrderStateException.php

Co-Authored-By: jolelievre <jo.lelievre@gmail.com>

matthieu-rolland committed:
Update src/Core/Domain/Order/Exception/InvalidOrderStateException.php

Co-Authored-By: jolelievre <jo.lelievre@gmail.com>

matthieu-rolland committed:
remove useless class in cancel product type form

matthieu-rolland committed:
move partial refund html outside of return product condition

matthieu-rolland committed:
lint

matthieu-rolland committed:
cancel product security anotations

matthieu-rolland committed:
release useless data attribute

matthieu-rolland committed:
fix products template issue

matthieu-rolland committed:
change behat then text

matthieu-rolland committed:
use invalidOrderStateException code errors in handlers and behat tests

matthieu-rolland committed:
improbe behat test with max cancellable quantity

matthieu-rolland committed:
remove useless EmptyProductSelectionException, use InvalidCancelProductException instead

matthieu-rolland committed:
add error messages for each invalid order state exceptions in orderController

matthieu-rolland commented:
@jo, @sowbiba:

I addressed all your feedbacks, except the refactor and modifications inside **CancelOrderProductHandler.php**. 

For now the code of this handler is really close to the legacy code, there is definitely space for code refactoring, but while doing it, more than one hour in I realized it's less simple than I thought it would be and the feature freeze should happen as soon as possible.

As discussed with @matks I think we have something that is working and has a well tested behavior, we should send this PR to QA asap.

I'm ok to rewrite thoughtfully this handler though, in a separate PR, after this PR is merged. Especially since our behat tests mitigate the risks of behavioral regressions.

jolelievre commented on admin-dev/themes/new-theme/js/pages/order/OrderViewPageMap.js:
```diff
@@ -169,6 +171,10 @@ export default {
+      cancelProducts: '.cancel-product-element:not(.hidden):not(.shipping-refund), .cancel-product-selector',
+    },
+    containers: {
+      refundCheckboxes: '.refund-checkboxes-container',
```
This selector seems useless now (not used in the js anymore), but I believe you still need to add some of the checkboxes If you need to hide all of them, based on what I explained above I believe you can simply add `:not(.refund-checkboxes-container)` in the selector so that the whole div will be hidden

matthieu-rolland replied:
removed

jolelievre commented on admin-dev/themes/new-theme/js/pages/order/view/order-product-cancel.js:
```diff
@@ -209,4 +209,15 @@ export default class OrderProductCancel {
+    this.initForm(
+        $(OrderViewPageMap.cancelProduct.buttons.save).data('cancelLabel'),
+        cancelProductRoute,
+        'cancel-product',
```
In `initForm` you need to add `cancel-product` to the removed class names or it will always be present if we switch from one mode to another (even though it shouldn't be possible)

matthieu-rolland replied:
done

jolelievre commented on src/Adapter/Order/CommandHandler/AbstractOrderCommandHandler.php:
```diff
@@ -63,8 +63,8 @@ protected function reinjectQuantity(OrderDetail $orderDetail, $productQuantity,
-
-        $orderDetail->product_quantity_reinjected += $quantityToReinject;
+        // @todo to be moved in OrderDetailUpdater service (introduced in migrate standard refund PR)
+        // $orderDetail->product_quantity_reinjected = $quantityToReinject;
```
To be moved in OrderRefundUpdater, along with a parameter `$reinjectProducts` or `$restockProducts`

matthieu-rolland replied:
done

jolelievre commented on src/Core/Form/IdentifiableObject/DataHandler/CancellationFormDataHandler.php:
```diff
@@ -0,0 +1,84 @@
+        $cancelledProducts = [];
+        foreach ($data['products'] as $product) {
+            if ($data['selected_' . $product->getOrderDetailId()]) {
+                $cancelledProducts[$product->getOrderDetailId()] = $data['quantity_' . $product->getOrderDetailId()];
```
Use `$data['quantity_' . $product->getOrderDetailId()] ?? 0` in case the field is not in the array

matthieu-rolland replied:
I don't think I should do that, I don't want to set the value to 0 in any case, I want the **$cancelledProducts** variable to contain selected products only

jolelievre replied:
I do it in other commands like `PartialRefundFormDataHandler` It will throw an exception soon after because the value is not allowed indeed (same for you in cancel command I believe)
This is better than a fatal error related to a non existent field in an array, at least you fully control the way it fails

matthieu-rolland replied:
I tried this approach, but it doesn't work in my case, if I apply your suggestion, then I can't distinguish between a product that was not selected, and a product that was selected but with a quantity of 0. I only want to throw an error in the last case.

(edit: and there is no fatal error, since the element is always in $data, it's a boolean. I added an **isset** to the condition, just for the sake of precaution though)

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:
```diff
@@ -1314,6 +1314,43 @@ public function sendProcessOrderEmailAction(Request $request): JsonResponse
 
+    /**
+     * @AdminSecurity(
+     *     "is_granted('update', request.get('_legacy_controller'))",
```
In legacy the required permission was `delete` Although it feels weird it also makes sense for this particular feature ^^ Maybe we should have them both? (same thing on return products and standard refund controllers, maybe you should update them as well?)
@MatShir what do you think?

matthieu-rolland replied:
I would say, from authorizations perspective, that this is an update of an order. We don't actually delete anything, we don't delete products, only relationships between products and orders, but that's really an update.

jolelievre replied:
I know, but sadly the legacy controller used this authorization: https://github.com/PrestaShop/PrestaShop/blob/68a6a0e8489edc57df4a097a656416f1fa52e440/controllers/admin/AdminOrdersController.php#L898

So I think we need to keep it (in addition with update which seems more logical)

matthieu-rolland replied:
alright, done !

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:
```diff
@@ -1365,29 +1402,29 @@ private function getErrorMessages(Exception $e)
-                    'Please enter a positive quantity to proceed with your refund.',
+            InvalidCancelProductException::class => [
+                InvalidCancelProductException::INVALID_QUANTITY => $this->trans(
+                    'Please enter a positive quantity',
```
I believe we still need a dot at the end
@LouiseBonnard what do you think? We simplified these messages because they are used in multiple features (cancel and refund)

LouiseBonnard replied:
Regarding final dots, there is no specific rule. Let's add it, but to all the error notifications this PR contains!

matthieu-rolland replied:
yep, I did exactly that earlier today 👍

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:
```diff
@@ -1396,6 +1433,24 @@ private function getErrorMessages(Exception $e)
             ),
+            InvalidOrderStateException::class => [
+                InvalidOrderStateException::UNEXPECTED_INVOICE => $this->trans(
+                    'Can not proceed on an order that already has an invoice.',
```
Suggestion: `Invalid action on an order that already has an invoice.`
To be confirmed by @LouiseBonnard (these messages can also be used for multiple contexts, we know the reason of the error but not always the initial context)

LouiseBonnard replied:
```suggestion
                    'Invalid action: this order already has an invoice.',
```

matthieu-rolland replied:
thanks @louise, so I use the same format for error messages just below ?

Ex: **Invalid action: this order not been delivered.**

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/OrderRefundFeatureContext.php:
```diff
@@ -223,18 +224,23 @@ public function enabledReturnProduct()
 
     /**
      * @Then I should get error that refund quantity is invalid
+     * @then I should get error that cancel quantity is invalid
```
```suggestion
     * @Then I should get error that cancel quantity is invalid
```

matthieu-rolland replied:
done

jolelievre reviewed (changes requested):
Still a few feedback but almost done

matthieu-rolland committed:
Properly deal with restock and reinjections in refunds

jolelievre commented on src/Adapter/Order/CommandHandler/IssuePartialRefundHandler.php:
```diff
@@ -116,10 +116,12 @@ public function handle(IssuePartialRefundCommand $command): void
 
         // @todo This part should probably be in a share abstract class as it will probably be common with other handlers
         // Update order details and reinject quantities
+        $shouldReinjectProducts = false;
```
```suggestion
        $shouldReinjectProducts = !$order->hasBeenDelivered() || $command->restockRefundedProducts();
```

matthieu-rolland replied:
done

matthieu-rolland committed:
address some feedbacks

matthieu-rolland committed:
change admin security grants

matthieu-rolland committed:
feedback

matthieu-rolland committed:
improve error messages

matthieu-rolland committed:
remove delete cart rule form and make it a link

matthieu-rolland committed:
style delete buttons

matthieu-rolland committed:
cs fixer

jolelievre reviewed (dismissed):
Great job @matthieu-rolland

LouiseBonnard commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:
```diff
@@ -1365,37 +1402,55 @@ private function getErrorMessages(Exception $e)
             ProductOutOfStockException::class => $this->trans(
-                'There are not enough products in stock',
+                'There are not enough products in stock.',
                 'Admin.Notifications.Error'
```
```suggestion
                'Admin.Catalog.Notification'
```

matthieu-rolland committed:
Update src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php

Co-Authored-By: LouiseBonnard <32565890+LouiseBonnard@users.noreply.github.com>

matthieu-rolland committed:
fix minor error message issue

matthieu-rolland committed:
use injected translator

matthieu-rolland committed:
hide shipping refund field in cancel mode

Robin-Fischer-PS commented:
Hi !

Some returns on this PR : 
- When canceling product, the Shipping refund block is displayed, it should not (since Cancel does not refund anything) : 
![pr17149 shipping block](https://user-images.githubusercontent.com/55688534/74245168-de0dfd00-4ce2-11ea-9dd9-165ae3889ec1.png)
- When I cancel all products of an order, an Exception is displayed. The exception is a missing translation of an error we have in Legacy.
Exception in Migrated page : 
![pr17179 cancel all products](https://user-images.githubusercontent.com/55688534/74244409-a6eb1c00-4ce1-11ea-9fa7-dfb2cddb0138.png)
Error in Legacy : 
![pr17149 cancel all legacy](https://user-images.githubusercontent.com/55688534/74244704-21b43700-4ce2-11ea-9ddf-d6d2300911e8.png)
So in fact, we have here two bugs : 
- The missing translation
- The error when cancelling a product, but this was already in Legacy (177 and 176)
 
I also made a comment to @MatShir on the issue, because spec is missing an info on "Merchandise return".

eternoendless commented:
Thank you @matthieu-rolland
